Bottom: a2deb7cf0ffc083566b6db6d7cef91c79198134b
Top:    bbeb47cdfb4bb65f97b22025fa5e4e45cb68940f
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-01-03 16:23:03 +0800

Refresh of tun-add-missing-xdp-flush

---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index e367d631..9448305 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -180,6 +180,7 @@ struct tun_file {
 	struct list_head next;
 	struct tun_struct *detached;
 	struct skb_array tx_array;
+	bool xdp_flush_needed;
 };
 
 struct tun_flow_entry {
@@ -1518,6 +1519,7 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 			err = xdp_do_redirect(tun->dev, &xdp, xdp_prog);
 			if (err)
 				goto err_redirect;
+			tfile->xdp_flush_needed = true;
 			rcu_read_unlock();
 			return NULL;
 		case XDP_TX:
@@ -1836,6 +1838,11 @@ static ssize_t tun_chr_write_iter(struct kiocb *iocb, struct iov_iter *from)
 	result = tun_get_user(tun, tfile, NULL, from,
 			      file->f_flags & O_NONBLOCK, false);
 
+	if (tfile->xdp_flush_needed) {
+		tfile->xdp_flush_needed = false;
+		xdp_do_flush_map();
+	}
+
 	tun_put(tun);
 	return result;
 }
@@ -2133,6 +2140,12 @@ static int tun_sendmsg(struct socket *sock, struct msghdr *m, size_t total_len)
 	ret = tun_get_user(tun, tfile, m->msg_control, &m->msg_iter,
 			   m->msg_flags & MSG_DONTWAIT,
 			   m->msg_flags & MSG_MORE);
+
+	if (tfile->xdp_flush_needed && !(m->msg_flags & MSG_MORE)) {
+		tfile->xdp_flush_needed = false;
+		xdp_do_flush_map();
+	}
+
 	tun_put(tun);
 	return ret;
 }
@@ -2945,6 +2958,8 @@ static int tun_chr_open(struct inode *inode, struct file * file)
 
 	sock_set_flag(&tfile->sk, SOCK_ZEROCOPY);
 
+	tfile->xdp_flush_needed = false;
+
 	return 0;
 }
